### ISSUE TITLE
dual focal loss

### ISSUE THREAD

**Issue by damhurmuller** [CoinCheung/pytorch-loss#6]
What is the equivalent of **Adaptive class weight layer** in the original article? The output of this layer should be the input of the softmax here:  `pred = torch.softmax(logits, dim=1)` but none treatment were made for this input.

**Comment by CoinCheung** [comment]
Personally, I doubt that whether this ACW layer is really helpful. The paper claims that ACW layer is equevalent to the `alpha` in the focal loss, however, you  can find in the paper of `focal loss` that the result of training is not sensitive to this `alpha` value. Additionally, the ACW layer requires learning a weight matrix `W_hat` for the output logits of the classifier which would makes the original classifier from `logits = W x + b` to `logits = W_hat (W x + b)`. As far as I know, stacking two linear layers without non-linearity added between them is equevalent to using a single linear layer. Therefore, I did not pay too much attention on it. Maybe I was wrong, you can also try it on yourself, and if it proves that I am wrong, you can kindly tell me and I am happy to fix it.

**Comment by damhurmuller** [comment]
Really thanks for the fast reply!

Accondlyng to the paper:

> To avoid the weighting factor to be larger than necessary, scaling coefficient **alpha** is applied to
diminish the weight intensity.

Mentioned that is equivalent to the weight w_i in the Weighted Cross Entropy Loss (WCE). Futhermore, in the paper there are experiments to check efficiency with and without. So, the ACW layer fine-tune the loss to improve the accuracy and prove that adaptible **alpha** is better than the fixed **alpha** in the WCE.

![acw_ce](https://user-images.githubusercontent.com/20431255/79412341-3b3c5e00-7f73-11ea-93d6-357eccb4d853.png)

![acw_dfl](https://user-images.githubusercontent.com/20431255/79412346-3d9eb800-7f73-11ea-9319-dd37cbe8a7ba.png)

Reference: https://arxiv.org/pdf/1909.11932.pdf


Actually, I  more interested in the fact that **alpha** can be learnable during the training process and can be adaptively adjustable for each class instead of only positive (binary) class. I'm trying to use this concept of learnable parameters in my research as master student.

I'm still learning pytorch, so do you think that is simple to code that layer with the learnable parameter?
Again, thanks for your help! :)

**Comment by CoinCheung** [comment]
I do not think defining a pytorch layer with a learnable parameter is impossible, you may refer to the implementation of [amsoftmax](https://github.com/CoinCheung/pytorch-loss/blob/master/amsoftmax.py) which contains a learnable parameter.

**Comment by CoinCheung** [comment]
Hi, I have no idea if you are still interested in implementing acw layer, but I have just thought that maybe you can use a 1x1 depth-wise convolution layer following the logits, which should work in the same way as the acw layer.

**Comment by damhurmuller** [comment]
Hi @CoinCheung . Thanks for the suggest! Can you explain more why the 1x1 depth-wise should work in the same way?

I was trying to implement using tf2. But I'm not sure if it is correct. I'm still confused why only one weight multiplying the output of logits will give me the best **alpha** for each class. This is not very intuitive for me.

```
class ACW(tf.keras.layers.Layer):
  def __init__(self, units=10):
    super(ACW, self).__init__()
    self.units = units

  def build(self, input_shape):  
    self.w = self.add_weight(shape=(input_shape[-1], self.units),
                             initializer='random_uniform', 
                             trainable=True)
    
    self.b = self.add_weight(shape=(self.units,),
                             initializer='zeros',
                             trainable=True)

  def call(self, inputs):
    tf.print(inputs)
    tf.print(self.w)
    return tf.matmul(inputs, self.w) + self.b
```

**Comment by CoinCheung** [comment]
I have just thought of that, if I have not misunderstood what ACW does, it seems that when the logits output of the model is a [N, M] tensor(N samples and each sample has M output values as logits of each class), ACW provides a weight vector of shape [1, M] that should be multiplied into the logits of each sample as the learned remedy for the knowledge learned from the unbalanced dataset. For each row of the [N, M] logits tensor, the weight vector is multiplied to it elementwisely, which I thought should be similar behavior of depthwise convolutions. For example, your model structure might be like this: 
```python
class Model(nn.Module):

    def __init__(self):
        self.backbone = some_fcn()
        self.acw = nn.Conv2d(in_chan, out_chan, 1, 1, 0, groups=in_chan)
        self.criteria = dual_focal_loss()

    def forward(self, x):
        feat = self.backbone(x)
        avg_pool = torch.mean(feat, dim=(2,3), keepdim=True)
        logits = self.acw(avg_pool)
        if self.training:
            out = self.criteria(logits)
        else:
            out = avg_pool
        return out
```
This `nn.Conv2d` with groups as same as input channels should act similarly as the acw layer.

**Comment by damhurmuller** [comment]
Thanks @CoinCheung ! This really makes sense. Thanks for the help and the explanation!
